repository/provider_metric.py

```diff
@@ -0,0 +1,106 @@
+from dataclasses import dataclass
+from datetime import datetime
+from enum import Enum
+from typing import List
+
+import requests
+
+
+class ProviderTrafficDirection(Enum):
+    IN = 1
+    OUT = 2
+
+
+@dataclass
+class ProviderTraffic:
+    provider_type: str
+    external_id: str
+    data_usage_bytes: int
+    date_from: datetime
+    date_to: datetime
+    direction: ProviderTrafficDirection
+
+
+class ProviderMetricException(Exception):
+
+    def __init__(self, message: str):
+        self.message = message
+
+
+class ProviderMetric:
+
+    def data_usage(self, date_from: datetime, date_to: datetime) -> List[ProviderTraffic]:
+        pass
+
+    def provider_type(self) -> str:
+        pass
+
+
+class PrometheusOpenvpnProviderMetric(ProviderMetric):
+
+    def __init__(self, prom_url: str):
+        self._prom_url = prom_url
+
+    def data_usage(self, date_from: datetime, date_to: datetime) -> List[ProviderTraffic]:
+        resp = requests.get(f"{self._prom_url}/api/v1/query?"
+                            f"query=sum by (common_name) (delta(openvpn_server_client_received_bytes_total"
+                            f"[{(date_to - date_from).seconds}s]))")
+        if resp.status_code == 200:
+            provider_traffics = []
+            for result in resp.json()['data']['result']:
```

Comment:
Simple is obviously better, but I disagree that it's simple :)
This way has a bit more things you should keep an eye on than `reduce` because you are managing these things manually.

But it's not a big deal here actually and I won't insist on any way))

Buuut I suspect you are confusing "simple" and "easy" and maybe you found [this ](https://youtu.be/vkUTX1hruF8?t=587 )interesting or even [this](https://www.youtube.com/watch?v=SxdOUGdseq4).

Reply:
Great, thanks for videos 👍

Reply:
Great, thanks for videos 👍